src/ralph/backends/database/clickhouse.py

```diff
@@ -0,0 +1,343 @@
+    ):
+        """Instantiates the ClickHouse client.
+
+        Args:
+            host (str): ClickHouse server host to connect to.
+            port (int): ClickHouse server port to connect to.
+            database (str): ClickHouse database to connect to.
+            event_table_name (str): Table where events live.
+            client_options (dict): A dictionary of valid options for the ClickHouse
+                client connection.
+        """
+        if client_options is None:
+            client_options = {
+                "date_time_input_format": "best_effort",  # Allows RFC dates
+                "allow_experimental_object_type": 1,  # Allows JSON data type
+            }
+
+        self.host = host
+        self.port = port
+        self.database = database
+        self.event_table_name = event_table_name
+
+        self.client = clickhouse_connect.get_client(
+            host=self.host,
+            port=self.port,
+            database=self.database,
+            settings=client_options,
+        )
+
+    def status(self) -> DatabaseStatus:
+        """Checks ClickHouse connection status."""
+        try:
+            self.client.query("SELECT 1")
+        except ClickHouseError:
+            return DatabaseStatus.AWAY
+
+        return DatabaseStatus.OK
+
+    @enforce_query_checks
+    def get(self, query: ClickHouseQuery = None, chunk_size: int = 500):
+        """Gets table rows and yields them."""
+        fields = ",".join(query.return_fields) if query.return_fields else "event"
+
+        sql = f"SELECT {fields} FROM {self.event_table_name}"
+
+        if query.where_clause:
+            sql += f"  WHERE {query.where_clause}"
+
+        result = self.client.query(sql).named_results()
+
+        for statement in result:
+            yield statement
+
+    @staticmethod
+    def to_documents(
+        stream: Union[TextIO, List], ignore_errors: bool = False
+    ) -> Generator[dict, None, None]:
+        """Converts `stream` lines (one statement per line) to insert tuples."""
+        for line in stream:
+            statement = json.loads(line) if isinstance(line, str) else line
+
+            try:
+                insert = ClickHouseInsert(
+                    event_id=statement["id"], emission_time=statement["timestamp"]
+                )
+
+                document = (
+                    insert.event_id,
+                    insert.emission_time,
+                    statement,
+                    json.dumps(statement),
+                )
+
+                yield document
+            except (KeyError, ValidationError) as exc:
+                err = (
+                    "Statement has an invalid or missing id or "
+                    f"timestamp field: {statement}"
+                )
+                if ignore_errors:
+                    logger.warning(err)
+                    continue
+                raise BadFormatException(err) from exc
```

Comment:
I usually prefer scoping try/except blocks to a single statement:

```suggestion
            try:
                insert = ClickHouseInsert(
                    event_id=statement["id"], emission_time=statement["timestamp"]
                )
            except (KeyError, ValidationError) as exc:
                err = (
                    "Statement has an invalid or missing id or "
                    f"timestamp field: {statement}"
                )
                if ignore_errors:
                    logger.warning(err)
                    continue
                raise BadFormatException(err) from exc

            document = (
                insert.event_id,
                insert.emission_time,
                statement,
                json.dumps(statement),
            )

            yield document
```